File: 2021/src/main/java/de/jfriemel/aoc/days/Day06.java
package de.jfriemel.aoc.days;

import de.jfriemel.aoc.Day;

import java.util.Arrays;
import java.util.List;
import java.util.stream.LongStream;

public class Day06 implements Day {

    @Override
    public String part1(List<String> input) {
        return Long.toString(simulate(input, 80));
    }

    @Override
    public String part2(List<String> input) {
        return Long.toString(simulate(input, 256));
    }

    private long simulate(final List<String> input, final int days) {
        final List<Integer> initialLanternfish = Arrays.stream(input.get(0)
                                                       .split(","))
                                                       .map(Integer::parseInt)
                                                       .toList();
        long[] lanternCounts = {0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L};

        for (final int lanternfish : initialLanternfish) {
            lanternCounts[lanternfish]++;
        }

        for (int i = 0; i < days; i++) {
            lanternCounts = simulateOnce(lanternCounts);
        }

        return LongStream.of(lanternCounts).sum();
    }

    private long[] simulateOnce(final long[] previous) {
        final long[] next = {0L, 0L, 0L, 0L, 0L, 0L, previous[0], 0L, previous[0]};
        for (int i = 1; i < previous.length; i++) {
            next[i - 1] += previous[i];
        }
        return next;
    }

}

File: 2023/src/main/kotlin/de/jfriemel/aoc/days/Day06.kt
package de.jfriemel.aoc.days

import de.jfriemel.aoc.Day

object Day06 : Day {
    override fun part1(input: List<String>): String {
        val races = input.map { line ->
            Regex("\\d+").findAll(line).map { matchResult -> matchResult.value.toInt() }.toList()
        }
        return races[0].indices.map { raceIndex ->
            (1..<races[0][raceIndex]).count { buttonTime ->
                buttonTime * (races[0][raceIndex] - buttonTime) > races[1][raceIndex]
            }
        }.reduce { a, b -> a * b }.toString()
    }

    override fun part2(input: List<String>): String {
        val races = input.map { line ->
            line.substringAfter(":").replace(" ", "").toLong()
        }
        return (1..<races[0]).count { buttonTime ->
            (buttonTime * (races[0] - buttonTime)) > races[1]
        }.toString()
    }
}
